pandas/core/internals.py

```diff
@@ -4485,33 +4492,64 @@ def _interleaved_dtype(blocks):
     for x in blocks:
         counts[type(x)].append(x)
 
-    have_int = len(counts[IntBlock]) > 0
     have_bool = len(counts[BoolBlock]) > 0
     have_object = len(counts[ObjectBlock]) > 0
+    have_int = len(counts[IntBlock]) > 0
     have_float = len(counts[FloatBlock]) > 0
     have_complex = len(counts[ComplexBlock]) > 0
     have_dt64 = len(counts[DatetimeBlock]) > 0
     have_dt64_tz = len(counts[DatetimeTZBlock]) > 0
     have_td64 = len(counts[TimeDeltaBlock]) > 0
-    have_cat = len(counts[CategoricalBlock]) > 0
+    have_cat = len(counts[CategoricalBlock])
     # TODO: have_sparse is not used
     have_sparse = len(counts[SparseBlock]) > 0  # noqa
-    have_numeric = have_float or have_complex or have_int
-    has_non_numeric = have_dt64 or have_dt64_tz or have_td64 or have_cat
+    have_numeric = have_float + have_complex + have_int
+    have_dt = have_dt64 + have_dt64_tz
+    have_non_numeric = have_dt64 + have_dt64_tz + have_td64 + have_cat
+    have_non_dt = have_td64 + have_cat
+    have_mixed = bool(have_numeric) + bool(have_non_dt) + bool(have_dt)
 
     if (have_object or
-        (have_bool and
-         (have_numeric or have_dt64 or have_dt64_tz or have_td64)) or
-        (have_numeric and has_non_numeric) or have_cat or have_dt64 or
-            have_dt64_tz or have_td64):
+            (have_non_numeric > 1) or  # more than one type of non numeric
+            (have_bool and have_mixed) or  # mix of a numeric et non numeric
+            (have_mixed > 1) or  # mix of a numeric et non numeric
+            have_dt64_tz or
+            (have_cat > 1)):
         return np.dtype(object)
+    elif have_dt64:
+        return np.dtype("datetime64[ns]")
+    elif have_td64:
+        return np.dtype("timedelta64[ns]")
     elif have_bool:
-        return np.dtype(bool)
+        return np.dtype("bool")
+    elif have_cat:
+        # return blocks[0].get_values().dtype
+        # if we are mixing unsigned and signed, then return
+        # the next biggest int type (if we can)
+
+        dts = [b.get_values().dtype for b in counts[CategoricalBlock]]
+        lcd = _find_common_type(dts)
+        kinds = set([_dt.kind for _dt in dts])
+
+        if len(kinds) == 1:
+            return lcd
+
+        if lcd == 'uint64' or lcd == 'int64':
+            return np.dtype('int64')
+
+        # return 1 bigger on the itemsize if unsinged
+        if lcd.kind == 'u':
+            return np.dtype('int%s' % (lcd.itemsize * 8 * 2))
```

Comment:
overal, it would be nice for internal functions/methods to have even a minimal docstring or comments to know the input/output expected and the objective ... not easy to understand otherwise the "contract" of the function.
I have the feeling the _interleaved_dtype function is similar in purpose to the _find_common_type function but limited to numpy dtypes (vs pandas dtypes), is that correct ?